Added the ability to customize the folder color.

proEndreeper opened the pull request:
This pull request changes so that folders can be assigned one of 7 colors, which are as follows:

- VANILLA (default grey)
- RED
- ORANGE
- YELLOW
- GREEN
- BLUE
- PURPLE

This change also involved making so the preferences could have default values, which involved setting up TypeAdapters for novels and folders so that Gson would not auto fill the missing values with null, so that the defaults in Folder.kt and WebNovel.kt would function correctly. This does add a small bit of overhead to both development and application, as now the TypeAdapters will need to be modified anytime the Folder or WebNovel structure is updated, but it does allow for default values for missing values, since adding the color field to the Folder will not retroactively update any existing users preferences, and the app would otherwise crash because the color field is null. Alternatively defaults can just be defined wherever the fields are used, but I felt it would work better if they were defined in the class itself, but please let me know if you want it to be different.

I also moved most of the preferences casting logic to another class (PreferenceUtils.kt) so that it is easier to maintain should any additional TypeAdapters that need to be created in the future. I also added global hideKeyboard iterations, as it was defined a few times across Fragments, and should make it easier to maintain in the future.

Here's a few pictures of the changes:
![image](https://user-images.githubusercontent.com/1905293/170163271-bd33b45f-daa4-4a7d-a1aa-24e2df18efde.png)
![image](https://user-images.githubusercontent.com/1905293/170163313-b8459bc8-ed63-4226-a2cd-dbd77eb1707d.png)
![image](https://user-images.githubusercontent.com/1905293/170164223-bf568f77-4eb5-4e64-907c-abad510278f7.png)


I did casually test if folders from a previous version would still load properly, and they did, though I might look at adding unit tests to the project to ensure code does not regress.

Please let me know if you want me to change anything to better fit your coding style or preferences.

ahmad-hossain commented on app/src/main/java/com/github/godspeed010/weblib/LibraryFragment.kt:
```diff
@@ -151,18 +178,12 @@ class LibraryFragment : Fragment() {
         et.requestFocus()
-        val imm = context?.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?
-        imm?.toggleSoftInput(InputMethodManager.SHOW_FORCED, InputMethodManager.HIDE_IMPLICIT_ONLY)
+        hideKeyboard()
```
The code you removed was for showing the keyboard so that when the Dialog is opened, the keyboard will popup so the user can start typing. 

However, you replaced the code with `hideKeyboard()` which does the opposite. I assume this was a mistake because calling `hideKeyboard()` here would have no effect, as the keyboard was never opened.

You could add the function for showing the keyboard in **ContextExtensions.kt**

As mentioned in a previous comment, the `hideKeyboard()` currently does not hide the keyboard. Once you have added the code for opening the keyboard, you can try testing if `hideKeyboard()` will close the keyboard when it was opened through that method.

proEndreeper replied:
Oops, I guess I just glanced over it and since it looks very similar to the original hide keyboard code, I had just replaced it with the hide keyboard function.

ahmad-hossain commented on app/src/main/java/com/github/godspeed010/weblib/NovelsFragment.kt:
```diff
@@ -171,20 +171,14 @@ class NovelsFragment : Fragment() {
-    private fun closeKeyboard() {
-        val imm = context?.getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager?
-        imm?.toggleSoftInput(InputMethodManager.HIDE_IMPLICIT_ONLY, 0)
+        hideKeyboard()
```
Same as mentioned in a previous comment. Keyboard should be opened in `focusEditText(..)`, not hidden.

proEndreeper replied:
This was another case of me not reading the entire line of focusEditText, but this has also been migrated now to use showKeyboard.

ahmad-hossain commented on app/src/main/res/layout/popup_folder_name.xml:
```diff
@@ -26,4 +27,11 @@
             android:hint="@string/hint_folder_name" />
 
     </com.google.android.material.textfield.TextInputLayout>
+
```
Would be great to have a little space between the EditText and the RadioGroup.

```suggestion

    <Space
        android:layout_width="match_parent"
        android:layout_height="15dp" />
```

proEndreeper replied:
Space has been added.

ahmad-hossain commented on app/src/main/java/com/github/godspeed010/weblib/LibraryFragment.kt:
```diff
@@ -132,17 +151,25 @@ class LibraryFragment : Fragment() {
+        }
+
+        inputColor.setOnClickListener {
+            hideKeyboard()
```
I tried manually focusing on the EditText by clicking on it, then I clicked any radio button. The keyboard did not become hidden. I can think of 3 possible reasons.

- Listener needs to go on RadioButtons rather than RadioGroup. I haven't played with RadioGroups/Buttons much so I don't know the answer to this off the top of my head, although I would think it would work.
- Code for `hideKeyboard()` does not function as intended
- Code for `hideKeyboard()` will only work when the keyboard is **opened** programmatically, similar to how you have code for hiding it. It may be that the keyboard will only hide when it was opened using a similar method, rather than when the user manually opens the keyboard.

proEndreeper replied:
I am not actually sure why this is not doing anything, but upon second thought, it doesn't make too much sense to me to close the keyboard if a user clicks on a color, as it does not affect the keyboard in any way. I did test theories 2 and 3, but the hide keyboard function still was not working. Theory 2 was tested by using your original code, and Theory 3 was tested with the same code used for Theory 2.

ahmad-hossain commented on app/src/main/java/com/github/godspeed010/weblib/ContextExtensions.kt:
```diff
@@ -0,0 +1,20 @@
+    view?.let { activity?.hideKeyboard(it) }
+}
+
+fun Activity.hideKeyboard() {
+    hideKeyboard(currentFocus ?: View(this))
+}
```
This function never gets called. When hideKeyboard() is called, it goes from `Fragment.hideKeyboard()` to `Context.hideKeyboard()`

I used Log statements and the Debugger to check this.

proEndreeper replied:
This function was intended for if the keyboard ever needed to be hidden from an Activity, though after going through your code, I am realizing that it's not needed in this app since almost everything that handles the keyboard is in a fragment or context related. The code for the ContextExtensions.kt was based off https://stackoverflow.com/a/45857155, as I am not as familiar with Android development when it came to hiding the keyboard, and was having difficulties with the keyboard not closing when any of the color options were clicked earlier on.

ahmad-hossain commented on app/src/main/java/com/github/godspeed010/weblib/LibraryFragment.kt:
```diff
@@ -97,6 +90,23 @@ class LibraryFragment : Fragment() {
+                arrayOf(intArrayOf(android.R.attr.state_enabled)),
+                intArrayOf(ContextCompat.getColor(requireContext(), it.rgbId))
+            )
+            button.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
+            (button.layoutParams as LinearLayout.LayoutParams).weight = 1f
+            radioGroup.addView(button)
```
I noticed that when I rotated my phone into Landscape Mode, the RadioGroup was not centered and leaning towards the left of the screen. 

Just re-looked at the Portrait mode too, and that's also off-center, although not as noticeable.

Please make both of them centered.

proEndreeper replied:
This involved a bit more than I was expecting, though I realized later what was causing it to have the additional spacing after it, for some reason having a weight different than 0 made a small space to the right of each radio button. I also ended up migrating the FolderColor radiogroup into a custom LinearLayout to simplify the interface.

ahmad-hossain commented on app/src/main/java/com/github/godspeed010/weblib/LibraryFragment.kt:
```diff
@@ -118,11 +131,17 @@ class LibraryFragment : Fragment() {
             DialogInterface.OnClickListener { dialog, which ->
                 dialog.dismiss()
-                closeKeyboard()
+                hideKeyboard()
```
I tried commenting out the `hideKeyboard()`, and the keyboard closes. 

Maybe it was never needed in the first place? I'm sure I would have realized if that was the case though... 

Actually, after some more thought, perhaps it's only needed when the keyboard is opened programatically.

proEndreeper replied:
I looked into this, and it seemed like it is needed if the soft keyboard is forced open.

ahmad-hossain reviewed (changes requested):
Hi proEndreeper,

Thank you for your contribution! I really like the feature you are implementing!! I have looked through the PR and added comments. Please go over them and don't hesitate to reach out if you have any questions/concerts.

Lastly, I realized my Play Store version was ahead of my Github. Sorry about that, I recently pushed the latest code into the repo so GitHub now has the Play Store version. Please merge that into your code.

GodSpeed010 commented:
> Lastly, I realized my Play Store version was ahead of my Github. Sorry about that, I recently pushed the latest code into the repo so GitHub now has the Play Store version. Please merge that into your code.

I just pushed another commit which organizes the files into folders. Please make sure you get that change too!

proEndreeper committed:
Added the ability to customize the folder color.

proEndreeper commented:
I am finding that the hideKeyboard function I reimplemented doesn't seem to be working correctly, so I changed it to the original code, but still have it separated out into the context. I also replaced focusEditText with showKeyboard in ContextExtensions, which uses the original code as well, as I was having issues getting InputMethodManager.hideSoftInputFromWindow and InputMethodManager.showSoftInput to work

I also made a change to the enter button functionality on the folders so that it will only close the dialog if the color group was already visible, as the color group is not visible on all devices.

ahmad-hossain commented on app/src/main/java/com/github/godspeed010/weblib/ContextExtensions.kt:
```diff
@@ -0,0 +1,29 @@
+
+fun Context.hideKeyboard(view: View) {
+    val inputMethodManager = getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
+    //inputMethodManager.hideSoftInputFromWindow(view.windowToken, 0)
```
Since this code isn't needed, it can be removed.
```suggestion
```

ahmad-hossain commented on app/src/main/java/com/github/godspeed010/weblib/ContextExtensions.kt:
```diff
@@ -0,0 +1,29 @@
+fun Context.showKeyboard(view: View) {
+    val inputMethodManager = getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
+    view.requestFocus()
+    //inputMethodManager.showSoftInput(view, InputMethod.SHOW_EXPLICIT)
```
Remove commented, unused code
```suggestion
```

ahmad-hossain commented on app/src/main/java/com/github/godspeed010/weblib/FolderColorGroup.kt:
```diff
@@ -0,0 +1,60 @@
+package com.github.godspeed010.weblib
```
> This involved a bit more than I was expecting, though I realized later what was causing it to have the additional spacing after it, for some reason having a weight different than 0 made a small space to the right of each radio button. I also ended up migrating the FolderColor radiogroup into a custom LinearLayout to simplify the interface.

I tried playing around with it and saw that it would be difficult to easily make the RadioButtons spread out evenly like you mentioned.
However, it seems here you decided to just add some padding while keeping all the buttons centered. If you would like to go with that, I was able to get that working using the old code.

I think this would be better because you don't have to create a new file with a good amount more code

1. Add `android:gravity="center"` to **RadioGroup** in **popup_folder_name.xml**
2. In LibraryFragment, you can keep most of the previous code. Just set the width to **wrap_content**, remove the weight, and set the padding.
```  
    private fun addColorsToDialog(radioGroup: RadioGroup)
    {
        enumValues<FolderColor>().forEach {
            val button = RadioButton(context)
            button.apply {
                text = ""
                tag = it
                buttonTintList = ColorStateList(
                    arrayOf(intArrayOf(android.R.attr.state_enabled)),
                    intArrayOf(ContextCompat.getColor(requireContext(), it.rgbId))
                )
                layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
                setPadding(15, 0, 15, 0)
            }
            radioGroup.addView(button)
        }
        radioGroup.check(radioGroup.getChildAt(0).id)
    }
```

proEndreeper replied:
Honestly, I liked the new implementation somewhat better simply because to change the folder color or retrieve the current selected color, it was just a single variable for either. Some of the thought was that if the color system is changed later on, it'd be easier to change it in a separate file than find the function for it in the LibraryFragment code.

proEndreeper replied:
I have changed it so it uses your modified version of the code instead of a separate class, also I apologize for cases where I use Kotlin inefficiently, I had not touched Kotlin nor Android development in depth prior to making the changes necessary for the Folders to have colors.

ahmad-hossain replied:
No worries, the code looks great, and I'm still learning more all the time too! Thanks for making the changes

ahmad-hossain reviewed (changes requested):
Hi proEndreeper,

Thanks again for the changes. I made some small comments.
As always, feel free to contact me with any questions.

ahmad-hossain reviewed (approved):
Looks great, thank you for making all the changes!
